Traitlets

rmorshea opened the pull request:
Refactoring of Matplotlib with traitlets. See TODO for progress.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/matplotlib/matplotlib/4762)
<!-- Reviewable:end -->

efiring commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,231 @@
+        in_range = False
+        if value is None or value is False or value in ['none','']:
+            # Return transparent if no other default alpha was set
+            return (0.0, 0.0, 0.0, 1.0)
```
This is black and opaque, not transparent.  I think you want all zeros.

efiring commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,231 @@
+            # Return transparent if no other default alpha was set
+            return (0.0, 0.0, 0.0, 1.0)
+
+        if isinstance(value, float) and 0 <= value <= 1:
```
This will eventually need to be modified, I expect:
```python
In [4]: isinstance(np.zeros((1,), dtype=np.float32)[0], float)
False
```

rmorshea replied:
There's a pull request to fix some things with the color traittype that i
probably forgot to include.
On Wed, Jul 22, 2015 at 11:03 PM Eric Firing <notifications@github.com>
wrote:

> In lib/matplotlib/traitlets.py
> <https://github.com/matplotlib/matplotlib/pull/4762#discussion_r35292090>:
>
> > +
> > +    def _float_to_shade(self, value):
> > +        grade = value*255.0
> > +        return (grade,grade,grade)
> > +
> > +    def _int_to_shade(self, value):
> > +        grade = value/255.0
> > +        return (grade,grade,grade)
> > +
> > +    def validate(self, obj, value):
> > +        in_range = False
> > +        if value is None or value is False or value in ['none','']:
> > +            # Return transparent if no other default alpha was set
> > +            return (0.0, 0.0, 0.0, 1.0)
> > +
> > +        if isinstance(value, float) and 0 <= value <= 1:
>
> This will eventually need to be modified, I expect:
>
> In [4]: isinstance(np.zeros((1,), dtype=np.float32)[0], float)False
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762/files#r35292090>.
>

efiring commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,231 @@
+    metadata = {
+        'force_rgb': False,
+        'as_hex' : False,
+        'default_alpha' : 0.0,
```
I think you want this to be 1.0.  Think of "alpha" as "opacity", not "transparency".

efiring commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,231 @@
+        }
+    allow_none = True
+    info_text = 'float, int, tuple of float or int, or a hex string color'
+    default_value = (0.0,0.0,0.0,0.0)
```
Likewise, what you have here is completely transparent.

efiring commented:
General strategy question: do you expect to be able to get to the test-and-fix stage (4) with one or a few traits at a time, or do you have to get everything in place before you can make anything work?

tacaswell commented on lib/matplotlib/artist.py:
```diff
@@ -85,12 +89,26 @@ class Artist(object):
+
+    stale = Bool(True)
+
+    def _stale_changed(self):
```
This should only trigger on the `False` -> `True` transistion

rmorshea replied:
Good point. Missed that in the last branch too.

rmorshea commented:
@efiring I'm hoping to be able to do things in batches. Right now I'm working with `stale`, `transform_set`, `transform`. The only reason I'm not doing large batches is to make tracing bugs easier.

tacaswell commented:
I would start with something simpler, like color or zorder rather than
transform

On Thu, Jul 23, 2015, 2:44 PM Ryan Morshead <notifications@github.com>
wrote:

> @efiring <https://github.com/efiring> I'm hoping to be able to do things
> in batches. Right now I'm working with stale, transform_set, transform.
> The only reason I'm not doing large batches is to make tracing bugs easier.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-124207070>
> .
>

efiring commented:
On 2015/07/23 8:44 AM, Ryan Morshead wrote:
> @efiring <https://github.com/efiring> I'm hoping to be able to do things
> in batches. Right now I'm working with |stale|, |transform_set|,
> |transform|. The only reason I'm not doing large batches is to make
> tracing bugs easier.

Good.  I advocate small batches for exactly that reason: much easier to 
find and fix bugs.  Better to keep it simple while gaining experience. 
These are massively intrusive changes.  It would be nice to get a sense 
of what they involve, and what sorts of consequences they have, sooner 
rather than later.

Do you have any way of evaluating the performance consequences of these 
changes?

efiring commented:
On 2015/07/23 8:46 AM, Thomas A Caswell wrote:
> I would start with something simpler, like color or zorder rather than
> transform

Good idea, though I think that color is actually one of the most 
complicated parameters--maybe *the* most complicated.

WeatherGod commented:
One argument for doing "stale", "transform", etc first is that these are
typically not user-facing, and they have fairly well defined interactions,
yet they have major performance implications. I would think that changes
here would be less intrusive, and fairly obvious if something has gone
wrong.

On Thu, Jul 23, 2015 at 2:55 PM, Eric Firing <notifications@github.com>
wrote:

> On 2015/07/23 8:46 AM, Thomas A Caswell wrote:
> > I would start with something simpler, like color or zorder rather than
> > transform
>
> Good idea, though I think that color is actually one of the most
> complicated parameters--maybe *the* most complicated.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-124210635>
> .
>

rmorshea commented:
```
-- Master Results --
avg wallclock: 0.0145939850807
avg user: 0.014611467
avg fps: 6852.13801761

-- TRAITLET RESULTS --
avg wallclock: 0.0166525568962
avg user: 0.0166520915
avg fps: 6005.08382126

Active TraitTypes:
+ transform
+ transform_set
+ stale
+ axes

* averaged over 2000 iterations
```

efiring commented:
Thanks for the benchmark results.  About a 14% penalty so far, in this
test.  What kind of a plot was it?

On 8/2/15, Ryan Morshead <notifications@github.com> wrote:
> ```
> -- Master Results --
> avg wallclock: 0.0145939850807
> avg user: 0.014611467
> avg fps: 6852.13801761
>
> -- TRAITLET RESULTS --
> avg wallclock: 0.0166525568962
> avg user: 0.0166520915
> avg fps: 6005.08382126
>
> Active TraitTypes:
> + transform
> + transform_set
> + stale
> + axes
>
> * averaged over 2000 iterations
> ```
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-126982783

rmorshea commented:
test code was take directly from [`simple_plot_fps.py`](https://github.com/rmorshea/matplotlib/blob/traitlets/examples/pylab_examples/simple_plot_fps.py)

rmorshea commented:
Rebased with the commits to #4694 made by @AndreLobato for the color TraitType. Updated `origin/master` to `upstream/master` then rebased `origin/traitlets` on top of `origin/master`. Ran the begining of the test suite again and the first error I encountered occurred when writing to a png from the pdf backend:

```
File "/Users/RyanMorshead/Coding/GitHub/matplotlib/lib/matplotlib/backends/backend_pdf.py", line 1288, in _writePng
    _png.write_png(data, buffer)
ValueError: Buffer must be RGBA NxMx4 array
```

Haven't investigated yet, but any chance that's coming from the changes to `upstream/master`? I wouldn't really expect the TraitTypes to influence that sort of thing.

tacaswell commented:
@rmorshea If that is happening locally, then the problem is out-dated c-cextensions (there was some recent work to improve the pdf backends handling of raster data).

rmorshea commented:
Is that something that requires work on your end to fix? Also, I'm presuming that Travis CI won't be able to build this branch until traitlets is officially supported?

mdboom commented:
Try deleting the build directory and rebuilding.  distutils doesn't do proper dependency resolution for C++ headers.

jenshnielsen commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,351 @@
+from .transforms import IdentityTransform, Transform
+import contextlib
+
+class exdict(dict):
```
Might want to inherit from userdict

rmorshea commented on lib/matplotlib/traitlets.py:
```diff
@@ -0,0 +1,374 @@
+            msg = "'%s' is not a trait of a %s class"
+            raise TraitError(msg % (name, self.__class__))
+        return trait
+
```
I'm having trouble with loading things after pickling. Right now, the `_default_generator` method for the "exdict" class I have for keeping track of old values is no longer there (as is expected since I removed it in `__getstate__`). I would expect the following bit of code in "PrivateMethodMixin" to fix this, however it instead raises an error claiming that `_rasterized` is no longer an attribute of a Rectangle instance. I suspect that this is because I'm overriding a `__setstate__` method somewhere else, but I'm not sure:

```python
    def __setstate__(self, state):
        meth = lambda klass, key: getattr(klass, key).default_value
        state['_trait_values']._default_generator = MethodType(meth, self.__class__)
        return state.update(self.__dict__)
```

rmorshea commented:
Failures in `traitlets` branch have been reduced to image comparisons in:
```
matplotlib.tests.test_path.test_xkcd.test
matplotlib.tests.test_text.test_basic_wrap.test
```
Both of these failures were also present the last time I ran the suite on `master`

rmorshea commented:
from `simple_traitlets_performance_test.py`:

```
-- Master Results --
avg fps: 7000 +/- 400

-- Traitlets Results --
avg fps: 6000 +/- 400

Active TraitTypes:
+ transform
+ transform_set
+ stale
+ axes
+ figure

* averaged over 2000 iterations
```

tacaswell commented:
Can you also compare against 1.4.3?

On Sat, Aug 15, 2015, 9:18 PM Ryan Morshead <notifications@github.com>
wrote:

> from simple_traitlets_performance_test.py:
>
> -- Master Results --
> avg fps: 7000 +/- 400
>
> -- Traitlets Results --
> avg fps: 6000 +/- 400
>
> Active TraitTypes:
> + transform
> + transform_set
> + stale
> + axes
> + figure
>
> * averaged over 2000 iterations
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-131473248>
> .
>

rmorshea commented:
```
-- v1.4.3 Results --
avg fps: ~7500

*averaged over 2000 iterations
```

rmorshea commented:
I ran these tests using `%matplotlib inline`.

rmorshea commented:
How is alpha blending implemented in the agg backend? The svg's seem to be fine, but the png comparison fails. I haven't actually changed anything in the agg backend while refactor the "alpha" attribute though, so I'm not sure what might be causing this.

rmorshea commented:
from the agg backend:
![test_alpha_blending_agg](https://cloud.githubusercontent.com/assets/4041990/9316925/1487161c-44ed-11e5-8a4d-7a0476d82558.png)
from the svg backend:
![test_alpha_blending_svg](https://cloud.githubusercontent.com/assets/4041990/9316926/148845f0-44ed-11e5-932f-82de262e1c92.png)

rmorshea commented:
@tacaswell I have some familiarity with cpp, so if the renderer really is the problem, I might be able investigate it given some direction as to where I should start looking in the renderer.

tacaswell commented:
@rmorshea I can't reproduce

![so](https://cloud.githubusercontent.com/assets/199813/9321104/0e464ca8-4530-11e5-9e3b-329ee4cd0445.png)

tacaswell commented:
Sorry, what I meant is that I can not reproduce this on master so the c++ renderer is probably ok.

rmorshea commented:
I'm just perplexed as to why one backend would work while another wouldn't.

rmorshea commented:
There's also a memory leak issue now in `test_quiver`, but I suspect that has something to do with this most recent rebase.

rmorshea commented:
is there an easy way to test the other backends for the same issue?

rmorshea commented:
Is there an easy way to switch backends to I can test the others?

tacaswell commented:
But, I can confirm that it is broken on this branch, let me see if I can make any progress on it.

tacaswell commented:
https://github.com/rmorshea/matplotlib/pull/7 fixes this issue.

rmorshea commented:
I've been thinking about how to meld the test suite with asv, and it would be nice to be able to integrate asv with nose so that any test failures could be evaluated during teardown rather simply throwing error which kills asv. However it doesn't seem that nose is included in the install of Python2.7 that asv uses. Is there anyone I should talk to about setting up nose in asv?

efiring commented:
What is "asv"?

rmorshea commented:
asv was brought up during one our recent google hangouts: https://github.com/spacetelescope/asv/

rmorshea commented:
@mdboom Just noticed that you've contributed to asv, do you have any suggestions on how asv and test suite might be integrated?

WeatherGod commented:
For those who don't want to bother with clicking links: asv == Air Speed
Velocity. Kinda like unit tests, but doesn't test correctness, only how
long they take to complete. It also builds up a history of these timings so
you can view a dashboard of results and see if how different parts of your
project is getting worse or better, performance-wise.

On Thu, Aug 20, 2015 at 8:19 PM, Ryan Morshead <notifications@github.com>
wrote:

> @mdboom <https://github.com/mdboom> Just noticed that you've contributed
> to asv, do you have any suggestions on how asv and test suite might be
> integrated?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-133225343>
> .
>

rmorshea commented:
Only failure is with the xkcd plots. Will be investigating that shortly. Beyond that, I plan to refactor the documentation generated by the examples.

jenshnielsen commented:
Im seeing a failure locally with one of the xkcd plots on master. The result and expected images are just a little bit different. Before spending to much time on it you should make sure that you don't have the failure on master

mdboom commented:
> do you have any suggestions on how asv and matplotlib's test suite might be integrated?

asv isn't really designed to integrate with a test suite -- just complement it.  I would suggest just starting with some simple benchmarks in asv and then add some that change properties dynamically (i.e. things that run through the traitlets code path).  Then we can see how we compare before and after...

rmorshea commented:
@jenshnielsen I was seeing that failure with xkcd on master as well.

rmorshea commented:
Finally found the memory leaks! Two in the new matplotlib specific traitlet work and two in the new traitlet api.

ellisonbg commented:
Great!

On Thu, Oct 1, 2015 at 11:14 PM, Ryan Morshead <notifications@github.com>
wrote:

> Finally found the memory leaks! Two in the new matplotlib specific
> traitlet work and two in the new traitlet api.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-144905037>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

rmorshea commented:
Only check failing is building the documentation and that's simply due to the warnings that are being raised. I'm now looking into how to rework `pchanged` and `stale` using traitlets. `pchanged` seems simpler that `stale` so I've started with that.

I'm running the test suite on something I mocked up for `pchanged` and it looks promising, however I'm getting some strange warnings. I'm thinking that may be due to the fact that the order in which `pchanged` and `stale` are triggered is significant, but I'm not sure. Could that be possible?

rmorshea commented:
I ran the test suite on master and realized that most of the errors I was seeing when I reworked `pchanged` and `stale` with traitlets were also occurring there. I haven't put my `pchanged` and `stale` implementations together yet but "traitletizing" them in this PR may be feasible.

mdboom commented:
@rmorshea, @ellisonbg: Finally able to come up for air and look at this in earnest.  What version of traitlets do I need to run this?  I'm getting lots of "cannot import name" errors with both traitlets 4.0.0 and git master.  I have some bandwidth to work on making things work together now, but thought I'd try to start from a known working state first.

rmorshea commented:
@mdboom at the moment, the last bit of decorator work in traitlets hasn't gone in yet (I presume the cannot imports are for `default`). So you'll have to use my traitlets branch until then:

https://github.com/rmorshea/traitlets/tree/default_decorators

rmorshea commented:
I'm noticing that there's something failing with respect to `transform_set` that got lost in all the traitlet warning messages I was suppressing. I'll try to have that fixed by the end of today.

mdboom commented:
Thanks, @rmorshea.  I also know you're back to other things, so I won't take offense if you just give me pointers to my questions rather than full solutions.

ellisonbg commented:
Mike,

We are about to release a new version of traitlets that has all of the new
APIs.

Cheers,

Brian

On Mon, Nov 2, 2015 at 3:45 PM, Michael Droettboom <notifications@github.com
> wrote:

> Thanks, @rmorshea <https://github.com/rmorshea>. I also know you're back
> to other things, so I won't take offense if you just give me pointers to my
> questions rather than full solutions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-153151145>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

rmorshea commented:
@mdboom I made commits to my branch of Traitlets and to this Mpl branch that should fix most issues, so if you didn't check out the branches within the last few hours, you should update them. The only thing wrong with my local build is in `lib/mpl_toolkits/axes_grid1/mpl_axes.py` that I think is stems from how default values are instantiated with my traitlets branch, but I haven't really looked into that.

ellisonbg commented:
Ryan. I talked with Sylvain today about the handling of @default stuff. I
think I am in favor of the decorator just defining the corresponding
_foo_default method on the class underneath so that the way it actually
works at runtime is the same.

On Mon, Nov 2, 2015 at 8:20 PM, Ryan Morshead <notifications@github.com>
wrote:

> @mdboom <https://github.com/mdboom> I made commits to my branch of
> Traitlets and to this Mpl branch that should fix most issues, so if you
> didn't check out the branches within the last few hours, you should update
> them. The only thing wrong with my local build is in
> lib/mpl_toolkits/axes_grid1/mpl_axes.py that I think is stems from how
> default values are instantiated with my traitlets branch, but I haven't
> really looked into that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-153210559>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

rmorshea commented:
@ellisonbg This is more of a discussion on a traitlets thread, but I'm a little worried that this sort of approach would be handled in the metaclass. This presumes that you would still allow for functions to be defined with names other than `_foo_default` though, where they would be renamed under the hood. Is that the case?

ellisonbg commented:
Yes, exactly, the decorator would simply be a fancy way of doing that
renaming and making it explicit and less magical.

On Mon, Nov 2, 2015 at 8:34 PM, Ryan Morshead <notifications@github.com>
wrote:

> @ellisonbg <https://github.com/ellisonbg> This is more of a discussion on
> a traitlets thread, but I'm a little worried that this sort of approach
> would be handled in the metaclass. This presumes that you would still allow
> for functions to be defined with names other than _foo_default though,
> where they would be renamed under the hood. Is that the case?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-153212076>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

mdboom commented:
How would I start playing with the serialization?  `Artist` (which is a subclass of `Configurable`) seems to always return an empty dictionary...

rmorshea commented:
```python
class A(Configurable):
	i = Int().tag(serialize=True)
	j = Int().tag(serialize=True)

	# this function doesn't exist (maybe it should).
	def trait_values(self, **metadata):
		values = dict()
		for n in self.trait_names(**metadata)
			values[n] = getattr(self, n)
		return values

serialized = self.trait_values(serialize=True)
```

The `Configurable` class (if I'm correct) is really just used to setup the default values for an instance using the `Config` of its parent or the one provided to it in `__init__`.

rmorshea commented:
I think the `@default` decorator will be merged soon. I'll let you know when you can drop my branch.

rmorshea commented:
@mdboom `@default` is now merged to master.

mdboom commented:
Ah -- so in short, we need to roll our own serialization (meaning a few lines of code to pick up all the traits recursively)?

ellisonbg commented:
Mike,

Here is a short summary of how I think about serialization with traitlets.

1. We probably want to start tagging the traits that should be serialized
using this API:

https://github.com/ipython/traitlets/blob/master/traitlets/tests/test_traitlets.py#L178

class Foo(HasTraits):
    a = Int(20).tag(serialize=True) # serialize this one
    b = Float(4.0) # not this one


2. We can then write serialization logic that walks the instance hierarchy
and queries each instance to find the traits that should be serialized:

f = Foo()
f.traits(serialize=True) # return the traits that should be serialized as a
dict

3. Write the serialization logic for each type.

There are two options here that make sense: 1) use something like multiple
dispatch to write a set of serialization function that are separate from
the classes and handle each type or 2) put to_dict methods on each class
and call things recursively.

Does this make sense?

Cheers,

Brian



On Thu, Nov 5, 2015 at 8:23 PM, Michael Droettboom <notifications@github.com
> wrote:

> Ah -- so in short, we need to roll our own serialization (meaning a few
> lines of code to pick up all the traits recursively)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4762#issuecomment-154255702>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

rmorshea commented:
I'm having difficulty tracking down this image comparison failure (which seems to manifest itself similarly in other image comparison failures involving polar plots and clipping):

**Result**
![polar_rmin_pdf](https://cloud.githubusercontent.com/assets/4041990/11167387/35bcf20c-8b15-11e5-9a26-94a2a827b74f.png)

**Expected**
![polar_rmin-expected](https://cloud.githubusercontent.com/assets/4041990/11167390/402b84ce-8b15-11e5-8650-d7ef2872ea5f.png)

**Test**
```python
r = np.arange(0, 3.0, 0.01)
theta = 2*np.pi*r

fig = plt.figure()
ax = fig.add_axes([0.1, 0.1, 0.8, 0.8], polar=True)
ax.plot(theta, r)
ax.set_rmax(2.0)
ax.set_rmin(0.5)
```

Obviously this has to do with the `viewLim` attribute (a `Bbox` instance) on the axes, however I'm not sure what's going wrong because using `set_rmax` updates the `viewLim` and runs `invalidate` (which I presume is the callback that informs parents of the change) without actually showing the change when it's rendered. Additionally this seems to be specific to polar plots since a similar test with cartesian axes properly clips extraneous data:

```python
fig = plt.figure()
ax = fig.add_axes([0,0,1,1])
ax.plot([1,2,3],[1,2,3])
ax.set_ylim(1,2)
```

I feel like I'm not looking in the right places; any thoughts on this problem?

rmorshea commented:
Additionally, I created and implemented a "batch" notifier (using the new decorator api) which is triggered when all the callbacks associated with a specified set of traits have been completed. This batch notifier may have other uses, but for now I've used it to clean up `pchanged` and `stale` in https://github.com/rmorshea/matplotlib/commit/32caa5f1650d1fe64b965589a3d28b5964fcb106. If it proves to be a useful idea here it might be something that could be pushed back into traitlets so feedback would be appreciated.

mdboom commented:
First, apologies to not getting back to you sooner.  I was traveling last week.

> I feel like I'm not looking in the right places -- any thoughts on this problem?

This looks related to the clip path -- polar plots have a circular clipping shape that is passed down to the renderer.  Does clip path have a trait?  Maybe somethings getting lost there.

rmorshea commented:
I thought it was the clip path as well, however on `master` the polar plot doesn't have a clip path. Rather the two plots have a `viewLim` attribute containing a `Bbox` instance that appears to contain this information. Using `set_rmax` will update the `Bbox`'s shape, however that change doesn't seem to be communicated when drawing. That's the part that I'm getting lost in.

tacaswell commented:
@QuLogic just overhauled how this worked (I think, I did not fully follow those changes).

rmorshea commented:
@tacaswell do you have any thoughts on of the "batch" notifier I mentioned above? In https://github.com/rmorshea/matplotlib/commit/32caa5f1650d1fe64b965589a3d28b5964fcb106 I used it to streamline how `stale` and `pchanged` are implemented.

tacaswell commented:
I have not had time to look at it enough to have thoughts :disappointed:

rmorshea commented:
I think I'm going to change the details of how it's implemented (and that would change more if it were pushed back to traitlets) but if you get the chance I'm interested in what you think of the idea in principle.

QuLogic commented:
No, not yet, the PR is still waiting on some bug fixes. Though when it's merged, you should get proper auto-refresh of the boundary patch.

mdboom commented:
@QuLogic: Which is the PR in question?

@rmorshea: artists on polar plots on master do have a clip path:

```
In [1]: import numpy as np

In [2]: import matplotlib.pyplot as plt

In [3]: r = np.arange(0, 3.0, 0.01)

In [4]: theta = 2 * np.pi * r

In [5]: ax = plt.subplot(111, projection='polar')

In [6]: line, = ax.plot(theta, r, color='r', linewidth=3)

In [7]: print(line.get_clip_path())
<matplotlib.transforms.TransformedPatchPath object at 0x7f664f67e710>
```

There is a bit of caching happening there -- it tries to only render the clip path once (which gets turned into a greyscale mask for rendering).  I'm going to investigate further with this branch.

mdboom commented:
I was able to fix the clip path issue with:

```diff
diff --git a/lib/matplotlib/artist.py b/lib/matplotlib/artist.py
index c610a7f..881802d 100644
--- a/lib/matplotlib/artist.py
+++ b/lib/matplotlib/artist.py
@@ -911,7 +911,7 @@ class Artist(PrivateMethodMixin, Configurable):
                 self.clippath = None
                 success = True
             elif isinstance(path, Patch):
-                self._clippath = TransformedPatchPath(path)
+                self.clippath = TransformedPatchPath(path)
                 success = True
             elif isinstance(path, tuple):
                 path, transform = path
@@ -923,14 +923,13 @@ class Artist(PrivateMethodMixin, Configurable):
             self.clippath = TransformedPath(path, transform)
             success = True
         elif isinstance(path, TransformedPatchPath):
-            self._clippath = path
+            self.clippath = path
             success = True
         elif isinstance(path, TransformedPath):
             self.clippath = path
             success = True
 
         if not success:
-            print(type(path), type(transform))
             raise TypeError("Invalid arguments to set_clip_path")
```

I think it's just the easy-to-miss `_` left in that should have been changed to the traitlet.

rmorshea commented:
I've hit a road bit of a road block:
+ I'm not sure where the error in python 2.7 is coming from, and I'm not able to replicate it on my machine.
+ Additionally the docs aren't building, though I haven't touched the docstring it's complaining about.

tacaswell commented:
I got the same sort of `getattr` error three times (once like travis and twice on a different test), but have not been able to reproduce it and get a debugger on it.  The failure on the scatter test probably is inducing the other 3 failures.

I can reproduce the sphinx issue locally, but no real guess at what is causing it.  I don't think it is something maformed with that intermediate output of the rst file, but I don't know where shpnix keeps them to check...

mdboom commented on lib/matplotlib/artist.py:
```diff
@@ -575,9 +802,14 @@ def get_sketch_params(self):
         May return `None` if no sketch parameters were set.
         """
-        return self._sketch
+        msg = ("This has been deprecated to make way for IPython's"
```
Maybe remove the `IPython's` here (and elsewhere similar messages appear).  This is not to avoid giving credit where credit is due, just that this could imply that we are growing a dependency on IPython, which we aren't.

mdboom commented:
Frustratingly, I can't reproduce the `getattr` error either.  What version of traitlets are @rmorshea or @tacaswell running?  Maybe that's relevant.

I'm looking into the Sphinx issue now...

rmorshea commented:
I'm running Traitlets off `master` right now.

tacaswell commented:
I copy-pasted the pip source install from .travis.yml.

rmorshea commented:
I'm going to finish rebasing (as best I can), and see where I'm at:

I feel like I've learned so much since I began this that it might be worth tearing everything down to start again with better practices and better ideas. If I had the time, I'd start from a fresh fork, but since I don't I'm hesitant.

I'll try to figure that out this week though.

@tacaswell

efiring commented:
@rmorshea my impression is that there have been some big changes in traitlets since you started.  Is this correct, and if so were they early enough in the process that you were able to take advantage of them?  In any case, I'm glad you are able to get back to this now.

efiring commented:
Answering my own  question, it looks like the major change was the introduction of decorators, which you are already using.  So the difficulties are because of all the mpl changes in the interim.

rmorshea commented:
@tacaswell unfortunately I'm putting this work on hold for about a month.

I am reviewing technical material for several job interviews and job fairs.

rmorshea commented:
@efiring, yes. The main outward change to Traitlets has been the new decorator API.

Current plan: rebuild from the ground up, but repurpose my past work wherever possible.

There are two reason that I want to do this:

1. There have been some internal changes to Traitlets that could allow me to build some custom tools for Matplotlib that will streamline certain validation, and notification pathways.
2. I can write cleaner, and more effective code than when I started this project.

rmorshea commented:
Rebase is unreasonably hairy - closing pending https://github.com/matplotlib/matplotlib/issues/8035's corresponding PR.

jklymak commented:
This can be revisited if someone takes up the traitlets torch

thisch commented:
> 3. There have been some internal changes to Traitlets that could allow me to build custom tools for Matplotlib that will streamline certain validation, and notification pathways.

@rmorshea Can you elaborate on this one?
